### DISCUSSION TITLE
How to add page

### DISCUSSION THREAD

**Discussion by kaaax0815** [sourcebroker/t3api#99] [Q&A] [answered]
I want to add a page using the api. How can I do that?
I am using the Mask extension

**Comment by klamparski** [comment, marked as answer]
@kaaax0815 There are two important informations you should start with:

1. t3api works on Extbase models, therefore firstly you would need to create a model and configure `pages` table to use it. It's quite easy with Extbase and it's well documented here: https://docs.typo3.org/m/typo3/reference-coreapi/main/en-us/ExtensionArchitecture/Extbase/Reference/Domain/Persistence.html#extbase_manual_mapping

2. Secondly you would need to define `POST` operation (endpoint) for your model. It's documented here: https://docs.typo3.org/p/sourcebroker/t3api/main/en-us/Operations/Index.html

You could use my example integration of `tt_content` and `pages` tables with t3api - [klamparski/t3apicontent](https://github.com/klamparski/t3apicontent). Although it's quite old and not updated, I believe implementation will be very similar in TYPO3 v12 and t3api 3.x.

**Reply by kaaax0815** [reply]
Thank you very much. Do you know how to make this work with the https://extensions.typo3.org/extension/mask extension?

**Reply by klamparski** [reply]
@kaaax0815 mask is just a solution to simplify creation of `tt_content` items. It means that it can be resolved in the same way as in my example above - [klamparski/t3apicontent](https://github.com/klamparski/t3apicontent). 

Specifically you can check [Configuration/Extbase/Persistence/Classes.php](https://github.com/klamparski/t3apicontent/blob/master/Configuration/Extbase/Persistence/Classes.php). There is a list of subclasses of `AbstractContentElement` to which entities from table `tt_content` are mapped depending on their value in `CType` column. You could do the same with ext:mask's content elements which simply have another value in this column.

**Reply by offizium-berndstorath** [reply]
@klamparski Thank you very much for your help.
I made a repo here: https://github.com/offizium-berndstorath/t3-page-api
If I use the `Web API` module and call `/_api/content/pages` I get 
```json
{
  "hydra:title": "The property \"pages\" on the subject does not exist.",
  "hydra:code": 1476109666,
  "hydra:description": "The property \"pages\" on the subject does not exist."
}
```
with `/_api/content/pages?page=1` I get
```json
{
  "reason": "Request parameters could not be validated (&cHash empty)"
}
```

**Reply by offizium-berndstorath** [reply]
@klamparski Sorry to bother you but Im lost here

**Reply by offizium-berndstorath** [reply]
@klamparski any update?

**Reply by klamparski** [reply]
@offizium-berndstorath I didn't have time yet to check it and I don't really know when I will be able to do so. I guess it would be better if you try to investigate it yourself, because I can't give you any date when I'll be able to check it mysefl.

**Reply by offizium-berndstorath** [reply]
@klamparski I did some research and can't find the cause of the error. I now switched to nnrestapi as it works
